### DISCUSSION TITLE
App construction failed (solved)

### DISCUSSION THREAD

**Discussion by gillsoftab** [uNetworking/uWebSockets.js#612] [General]
When building with nodejs 16.10.0 (nodejs 16.9.0+) on Debian 11 i get the error **Error: App construction failed** when trying to start Https server instance. Http works. However, when building for nodejs 16.8.0 everyhing works for Https too.

Is any changes needed in this repo, uWebSockets or uSockets in order to use nodejs 16.9.0+ with Https?
Does anyone else have the same problem?

**Comment by ghost** [comment]
https://github.com/uNetworking/uWebSockets.js/blob/master/src/AppWrapper.h#L527

It fails if you specify wrong SSL cert path, etc.

**Reply by gillsoftab** [reply]
I provide the exact same cert for nodejs 16.8.0 and nodejs 16.10.0. The only thing I change is the node version. 

gcc (Debian 10.2.1-6) 10.2.1 20210110

**Might be related:**
_I checked the function emplace_back and found it existed only in GCC 6.3 and deleted in 10.2.
I do not know why db2ia.node refer to this deleted function? Maybe we need to upgrade V8 or node API?_
https://github.com/IBM/nodejs-idb-connector/issues/141#issuecomment-828097520

**Comment by ghost** [comment]
I think you're overcomplicating this.

1) you shouldn't build the source, you should just install the binaries. They work on any official Node.js 16 and they follow official ABIs.
2) that runtime error is because SSL failed to init (most likely wrong cert path, but can also be because someone fucked up ABIs or linked incorrectly). Either case, we only support official Node.js binaries, using our compiled binaries.

What Node.js or IBM or Debian does on their part is not part of this project, we can only play by the official ABI rules.

**Reply by gillsoftab** [reply]
Ok, thanks.

Most likely nodejs developers didin't think all the way. They also have made some other error related to certificates so likely a nodejs issue. I'll have to stick to nodejs 16.8.0 until they fix their introduced bug.

**Reply by ghost** [reply]
Note that uWS shares nothing with Node.js. They are two entirely different non-sharing pieces of code. The only thing that is common is both link to OpenSSL, no other commonalities exist. So issues regarding Node.js certificates or whatnot are definitely not affecting uWS.

**Comment by ghost** [comment]
In fact, because of issues with Linux distros not following official ABIs, we will probably move to our own BoringSSL dependency shipped as part of uWS instead of linking with Node.js's OpenSSL. Then we can completely eliminate all ABI issues and be more compatible with obscure Linux distros. This is also needed for QUIC support and runs faster.

**Comment by gillsoftab** [comment]
Thanks, for the hint (i.e. SSL failed to init), I tracked down the issue from that (calling a client function). My build now works with nodejs 16.10.0.
